pkgs/hook/tool/generate.dart

```diff
@@ -0,0 +1,1270 @@
+      if (nameParts.length == 1 && nameParts.single.startsWithUpperCase()) {
+        return nameParts.single;
+      }
+    }
+    return null;
+  }
+
+  /// This is all the inferred class names from definitions in order of
+  /// traversal.
+  List<String> get classNames {
+    final result = <String>[];
+    for (final path in paths) {
+      final className = _pathToClassName(path);
+      if (className != null) {
+        if (!result.contains(className)) {
+          result.add(className);
+        }
+      }
+    }
+    return result;
+  }
+
+  String? get className => classNames.firstOrNull;
+
+  String? get superClassName {
+    final names = classNames;
+    if (names.length == 2) {
+      return names[1];
+    }
+    if (names.length > 2) {
+      throw UnimplementedError();
+    }
+    return null;
+  }
+
+  JsonSchemas? get superClassSchemas {
+    final parentClassName = superClassName;
+    if (parentClassName == null) {
+      return null;
+    }
+    for (final schema in _schemas) {
+      final path = schema.path;
+      if (path == null) continue;
+      final className = _pathToClassName(path);
+      if (className != parentClassName) continue;
+      return JsonSchemas(schema)._flatten();
+    }
+    throw UnimplementedError();
+  }
+
+  JsonSchemas get baseSchemas {
+    final packages = dartPackages;
+    if (packages.length == 1) {
+      return this;
+    }
+    if (packages.length != 2) {
+      throw UnimplementedError();
+    }
+    final package = dependency(packages);
+    final result = <JsonSchema>[];
+    for (final schema in _schemas) {
+      if (schema.dartPackage == package) {
+        result.add(schema);
+      }
+    }
+    return JsonSchemas._(result);
+  }
+
+  JsonSchemas get extensionSchemas {
+    final packages = dartPackages;
+    if (packages.length == 1) {
+      return JsonSchemas._([]);
+    }
+    if (packages.length != 2) {
+      throw UnimplementedError();
+    }
+    final package = importer(packages);
+    final result = <JsonSchema>[];
+    for (final schema in _schemas) {
+      if (schema.dartPackage == package) {
+        result.add(schema);
+      }
+    }
+    return JsonSchemas._(result);
+  }
+
+  bool generateExtension(String currentPackage) =>
+      dartPackages.length > 1 ||
+      dartPackages[0] != currentPackage && className != null;
+
+  List<String> get dartPackages {
+    final result = <String>[];
+    for (final schema in _schemas) {
+      final pkg = schema.dartPackage;
+      if (!result.contains(pkg)) {
+        result.add(pkg);
+      }
+    }
+    return result;
+  }
+}
+
+extension on JsonSchema {
+  String get dartPackage {
+    final schemaUri = rootSchemas2[root]!;
+    final segments = schemaUri.pathSegments;
+    final pkgName = segments[segments.indexOf('pkgs') + 1];
+    return pkgName;
+  }
+}
+
+extension on String {
+  bool startsWithUpperCase() {
+    final codeUnit = codeUnitAt(0);
+    return codeUnit >= 65 && codeUnit <= 90;
```

Comment:
nit: I'd prefer `this[0] == this[0].toUppercase()` or `'A'.codeUnitAt(0)` instead of the number.